## Segmenting and Clustering Neighbourhoods in Toronto

### Part 1: Converting Wikipedia html table into a DataFrame

In [34]:
import numpy as np
import pandas as pd
from pandas import DataFrame
from bs4 import BeautifulSoup as bsoup
from urllib.request import urlopen as uReq
import requests
import lxml

In [35]:
source_file='https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M'
r=requests.get(source_file)

#### 1.1 Read the web html file with BeautifulSoup 

In [36]:
page_content=bsoup(r.text,"html.parser")
#page_content

#### 1.2 Find Table in html file

In [37]:
mytable=page_content.table
#mytable

In [38]:
results=mytable.find_all('tr')
no_lines=len(results)
print('now we have found', no_lines, ' lines incl. the Header, which is the first line with 0 index')

now we have found 288  lines incl. the Header, which is the first line with 0 index


In [39]:
header=results[0].text.split()
header

['Postcode', 'Borough', 'Neighbourhood']

#### 1.3 Create a loop to extract the table contents into a dataframe df1

In [40]:
lines =[]
n=1
while n < no_lines :
    Postcode=results[n].text.split('\n')[1]
    Borough=results[n].text.split('\n')[2]
    Neighborhood=results[n].text.split('\n')[3]
    lines.append((Postcode, Borough,Neighborhood))
    n=n+1

df1=pd.DataFrame(lines, columns=['PostalCode', 'Borough', 'Neighbourhood'])
df1.head(5)

PostalCode           Borough     Neighbourhood
0        M1A      Not assigned      Not assigned
1        M2A      Not assigned      Not assigned
2        M3A        North York         Parkwoods
3        M4A        North York  Victoria Village
4        M5A  Downtown Toronto      Harbourfront

In [14]:
df1.shape

(287, 3)

#### 1.4 Remove all rows with Borough = 'Not assigned' 

In [41]:
df1[df1['Borough']=='Not assigned'].count()

PostalCode       77
Borough          77
Neighbourhood    77
dtype: int64

In [42]:
df1.drop(df1[df1.Borough =='Not assigned'].index, inplace = True)
df1.reset_index(drop=True, inplace = True)

In [43]:
df1.head(5)

PostalCode           Borough     Neighbourhood
0        M3A        North York         Parkwoods
1        M4A        North York  Victoria Village
2        M5A  Downtown Toronto      Harbourfront
3        M6A        North York  Lawrence Heights
4        M6A        North York    Lawrence Manor

In [18]:
df1.tail()

PostalCode    Borough             Neighbourhood
205        M8Z  Etobicoke  Kingsway Park South West
206        M8Z  Etobicoke                 Mimico NW
207        M8Z  Etobicoke        The Queensway West
208        M8Z  Etobicoke     Royal York South West
209        M8Z  Etobicoke            South of Bloor

In [44]:
df1.shape

(210, 3)

#### 1.5 Replace 'Not assigned' neighborhoods with the name of the Borough

In [20]:
df1[df1['Neighbourhood']=='Not assigned'].count()

PostalCode       1
Borough          1
Neighbourhood    1
dtype: int64

In [45]:
df1.loc[df1['Neighbourhood'] == 'Not assigned', 'Neighbourhood'] = df1['Borough']

In [46]:
df1[df1['Neighbourhood']=='Not assigned'].count()

PostalCode       0
Borough          0
Neighbourhood    0
dtype: int64

##### 1 row with 'not assigned' Neighborhood was replaced by Borough

In [47]:
df1.shape

(210, 3)

#### 1.6 Combine into one row all PostalCodes/ Boroughs with several neighborhoods with the neighborhoods separated with a comma

In [48]:
print('Target Data Frame should only consists of ', len(df1.groupby(['PostalCode','Borough'])), ' lines')


Target Data Frame should only consists of  103  lines


In [49]:
df1 = df1[~df1['Borough'].isnull()]  # to filter out bad rows

In [39]:
df1.shape

(210, 3)

In [50]:
df_target = df1.groupby(['PostalCode','Borough'])['Neighbourhood'].apply(lambda x: ','.join(x)).reset_index()

In [45]:
df_target.head(5)

PostalCode      Borough                         Neighbourhood
0        M1B  Scarborough                         Rouge,Malvern
1        M1C  Scarborough  Highland Creek,Rouge Hill,Port Union
2        M1E  Scarborough       Guildwood,Morningside,West Hill
3        M1G  Scarborough                                Woburn
4        M1H  Scarborough                             Cedarbrae

In [46]:
df_target.shape

(103, 3)

### Now we have the first part of Target Data Frame ready with only 103 lines

### End of Part 1 of 2 of Toronto Project --------------------------

### Part 2: Get the latitude and the longitude coordinates of each neighborhood

#### 2.1 Get the data from Geodata URL

In [51]:
url='http://cocl.us/Geospatial_data'
geodata_df=pd.read_csv(url)
geodata_df.head()

Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

In [49]:
geodata_df.shape

(103, 3)

In [52]:
geodata_df.rename(columns={'Postal Code':'PostalCode'},inplace=True)
geodata_df.head()

PostalCode   Latitude  Longitude
0        M1B  43.806686 -79.194353
1        M1C  43.784535 -79.160497
2        M1E  43.763573 -79.188711
3        M1G  43.770992 -79.216917
4        M1H  43.773136 -79.239476

#### 2.2 Merge the target Data Frame with Geodata

In [54]:
df_merged = df_target.join(geodata_df.set_index('PostalCode'), on='PostalCode')
df_merged.head(5)

PostalCode      Borough                         Neighbourhood   Latitude  \
0        M1B  Scarborough                         Rouge,Malvern  43.806686   
1        M1C  Scarborough  Highland Creek,Rouge Hill,Port Union  43.784535   
2        M1E  Scarborough       Guildwood,Morningside,West Hill  43.763573   
3        M1G  Scarborough                                Woburn  43.770992   
4        M1H  Scarborough                             Cedarbrae  43.773136   

   Longitude  
0 -79.194353  
1 -79.160497  
2 -79.188711  
3 -79.216917  
4 -79.239476

### End of Part 2 of 2 of Toronto Project ------------------------

### Part 3: Exploring Toronto Neighbourhood

#### 3.1 Select Data Frame with only Toronto boroughs

In [55]:
Toronto_data = df_merged[df_merged['Borough'].str.contains('Toronto', na = False)].reset_index(drop=True)
Toronto_data.head(5)

PostalCode          Borough                  Neighbourhood   Latitude  \
0        M4E     East Toronto                    The Beaches  43.676357   
1        M4K     East Toronto    The Danforth West,Riverdale  43.679557   
2        M4L     East Toronto  The Beaches West,India Bazaar  43.668999   
3        M4M     East Toronto                Studio District  43.659526   
4        M4N  Central Toronto                  Lawrence Park  43.728020   

   Longitude  
0 -79.293031  
1 -79.352188  
2 -79.315572  
3 -79.340923  
4 -79.388790

In [56]:
Toronto_data.shape


(38, 5)

#### Now we have 38 lines (37 without header) in Toronto Data Frame

In [32]:
Toronto_data.tail()

PostalCode       Borough  \
33        M6K  West Toronto   
34        M6P  West Toronto   
35        M6R  West Toronto   
36        M6S  West Toronto   
37        M7Y  East Toronto   

                                        Neighbourhood   Latitude  Longitude  
33         Brockton,Exhibition Place,Parkdale Village  43.636847 -79.428191  
34                       High Park,The Junction South  43.661608 -79.464763  
35                              Parkdale,Roncesvalles  43.648960 -79.456325  
36                                  Runnymede,Swansea  43.651571 -79.484450  
37  Business Reply Mail Processing Centre 969 Eastern  43.662744 -79.321558

#### 3.2 Create a map of Toronto

In [57]:
latitude = 43.651571
longitude= -79.484450

In [58]:
# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

#!conda install -c conda-forge folium=0.5.0 --yes # uncomment this line if you haven't completed the Foursquare API lab
import folium # map rendering library

In [59]:
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=11)

# add markers to map
for lat, lng, label in zip(Toronto_data['Latitude'], Toronto_data['Longitude'], Toronto_data['Neighbourhood']):
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

#### 3.3 Define Foursquare Credentials and Version

In [60]:
# The code was removed by Watson Studio for sharing.

In [62]:
CLIENT_ID = MyClientId # your Foursquare ID
CLIENT_SECRET = MyClientSecret # your Foursquare Secret
VERSION = '20191130' # Foursquare API version

#print('Your credentails:')
#print('CLIENT_ID: ' + CLIENT_ID)
#print('CLIENT_SECRET:' + CLIENT_SECRET)

In [65]:
print('1: ' , Toronto_data.loc[0, 'Neighbourhood'])
print('2: ' , Toronto_data.loc[1, 'Neighbourhood'])
print('3: ' , Toronto_data.loc[2, 'Neighbourhood'])
print('4: ' , Toronto_data.loc[3, 'Neighbourhood'])
print('5: ' , Toronto_data.loc[5, 'Neighbourhood'])

1:  The Beaches
2:  The Danforth West,Riverdale
3:  The Beaches West,India Bazaar
4:  Studio District
5:  Davisville North


In [67]:
neighbourhood_latitude = Toronto_data.loc[0, 'Latitude'] # neighbourhood latitude value
neighbourhood_longitude = Toronto_data.loc[0, 'Longitude'] # neighbourhood longitude value

neighbourhood_name = Toronto_data.loc[0, 'Neighbourhood'] # neighbourhood name

print('Latitude and longitude values of {} are {}, {}.'.format(neighbourhood_name, 
                                                               neighbourhood_latitude, 
                                                               neighbourhood_longitude))

Latitude and longitude values of The Beaches are 43.67635739999999, -79.2930312.


#### 3.4 Create a function to repeat the same process to all the neighborhoods in Toronto Dataframe

In [68]:
LIMIT = 100 # limit of number of venues returned by Foursquare API
radius = 500 # define radius

In [69]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)


In [73]:
toronto_venues = getNearbyVenues(names=Toronto_data['Neighbourhood'],
                                   latitudes=Toronto_data['Latitude'],
                                   longitudes=Toronto_data['Longitude']
                                  )

The Beaches
The Danforth West,Riverdale
The Beaches West,India Bazaar
Studio District
Lawrence Park
Davisville North
North Toronto West
Davisville
Moore Park,Summerhill East
Deer Park,Forest Hill SE,Rathnelly,South Hill,Summerhill West
Rosedale
Cabbagetown,St. James Town
Church and Wellesley
Harbourfront
Ryerson,Garden District
St. James Town
Berczy Park
Central Bay Street
Adelaide,King,Richmond
Harbourfront East,Toronto Islands,Union Station
Design Exchange,Toronto Dominion Centre
Commerce Court,Victoria Hotel
Roselawn
Forest Hill North,Forest Hill West
The Annex,North Midtown,Yorkville
Harbord,University of Toronto
Chinatown,Grange Park,Kensington Market
CN Tower,Bathurst Quay,Island airport,Harbourfront West,King and Spadina,Railway Lands,South Niagara
Stn A PO Boxes 25 The Esplanade
First Canadian Place,Underground city
Christie
Dovercourt Village,Dufferin
Little Portugal,Trinity
Brockton,Exhibition Place,Parkdale Village
High Park,The Junction South
Parkdale,Roncesvalles
Runnymede

##### let's check the size of the resulting dataframe

In [74]:
print(toronto_venues.shape)
toronto_venues.head()

(1685, 7)


Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0                  The Beaches              43.676357              -79.293031   
1                  The Beaches              43.676357              -79.293031   
2                  The Beaches              43.676357              -79.293031   
3                  The Beaches              43.676357              -79.293031   
4  The Danforth West,Riverdale              43.679557              -79.352188   

                                Venue  Venue Latitude  Venue Longitude  \
0                   Glen Manor Ravine       43.676821       -79.293942   
1  The Big Carrot Natural Food Market       43.678879       -79.297734   
2                 Grover Pub and Grub       43.679181       -79.297215   
3                       Upper Beaches       43.680563       -79.292869   
4                            Pantheon       43.677621       -79.351434   

      Venue Category  
0              Trail  
1  Health Food Store  
2                Pub  
3       Neighborhood  
4   Greek Restaurant

##### Let's find out how many unique categories can be curated from all the returned venues

In [75]:
print('There are {} uniques categories.'.format(len(toronto_venues['Venue Category'].unique())))


There are 235 uniques categories.


In [76]:
toronto_venues.groupby('Neighborhood').count()

Neighborhood Latitude  \
Neighborhood                                                                
Adelaide,King,Richmond                                                100   
Berczy Park                                                            57   
Brockton,Exhibition Place,Parkdale Village                             21   
Business Reply Mail Processing Centre 969 Eastern                      17   
CN Tower,Bathurst Quay,Island airport,Harbourfr...                     14   
Cabbagetown,St. James Town                                             46   
Central Bay Street                                                     80   
Chinatown,Grange Park,Kensington Market                                93   
Christie                                                               17   
Church and Wellesley                                                   84   
Commerce Court,Victoria Hotel                                         100   
Davisville                                                             36   
Davisville North                                                        8   
Deer Park,Forest Hill SE,Rathnelly,South Hill,S...                     15   
Design Exchange,Toronto Dominion Centre                               100   
Dovercourt Village,Dufferin                                            18   
First Canadian Place,Underground city                                 100   
Forest Hill North,Forest Hill West                                      4   
Harbord,University of Toronto                                          35   
Harbourfront                                                           50   
Harbourfront East,Toronto Islands,Union Station                       100   
High Park,The Junction South                                           24   
Lawrence Park                                                           3   
Little Portugal,Trinity                                                62   
Moore Park,Summerhill East                                              3   
North Toronto West                                                     20   
Parkdale,Roncesvalles                                                  14   
Rosedale                                                                4   
Roselawn                                                                1   
Runnymede,Swansea                                                      36   
Ryerson,Garden District                                               100   
St. James Town                                                        100   
Stn A PO Boxes 25 The Esplanade                                        98   
Studio District                                                        37   
The Annex,North Midtown,Yorkville                                      21   
The Beaches                                                             4   
The Beaches West,India Bazaar                                          20   
The Danforth West,Riverdale                                            43   

                                                    Neighborhood Longitude  \
Neighborhood                                                                 
Adelaide,King,Richmond                                                 100   
Berczy Park                                                             57   
Brockton,Exhibition Place,Parkdale Village                              21   
Business Reply Mail Processing Centre 969 Eastern                       17   
CN Tower,Bathurst Quay,Island airport,Harbourfr...                      14   
Cabbagetown,St. James Town                                              46   
Central Bay Street                                                      80   
Chinatown,Grange Park,Kensington Market                                 93   
Christie                                                                17   
Church and Wellesley                                                    84   
Commerce Court,Victoria Hotel      

#### 3.5 Analyze Each Neighborhood

In [77]:
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
toronto_onehot['Neighborhood'] = toronto_venues['Neighborhood'] 

# move neighborhood column to the first column
cols=list(toronto_onehot.columns.values)
cols.pop(cols.index('Neighborhood'))
toronto_onehot=toronto_onehot[['Neighborhood']+cols]

# rename Neighborhood for Neighbourhood so that future merge works
toronto_onehot.rename(columns = {'Neighborhood': 'Neighbourhood'}, inplace = True)
toronto_onehot.head()

Neighbourhood  Afghan Restaurant  Airport  \
0                  The Beaches                  0        0   
1                  The Beaches                  0        0   
2                  The Beaches                  0        0   
3                  The Beaches                  0        0   
4  The Danforth West,Riverdale                  0        0   

   Airport Food Court  Airport Lounge  Airport Service  Airport Terminal  \
0                   0               0                0                 0   
1                   0               0                0                 0   
2                   0               0                0                 0   
3                   0               0                0                 0   
4                   0               0                0                 0   

   American Restaurant  Antique Shop  Aquarium  ...  Toy / Game Store  Trail  \
0                    0             0         0  ...                 0      1   
1                    0             0         0  ...                 0      0   
2                    0             0         0  ...                 0      0   
3                    0             0         0  ...                 0      0   
4                    0             0         0  ...                 0      0   

   Train Station  Vegetarian / Vegan Restaurant  Video Game Store  \
0              0                              0                 0   
1              0                              0                 0   
2              0                              0                 0   
3              0                              0                 0   
4              0                              0                 0   

   Vietnamese Restaurant  Wine Bar  Wine Shop  Wings Joint  Yoga Studio  
0                      0         0          0            0            0  
1                      0         0          0            0            0  
2                      0         0          0            0            0  
3                      0         0          0            0            0  
4                      0         0          0            0            0  

[5 rows x 235 columns]

##### Next, let's group rows by neighborhood and by taking the mean of the frequency of occurrence of each category

In [78]:
toronto_grouped = toronto_onehot.groupby('Neighbourhood').mean().reset_index()
toronto_grouped

Neighbourhood  Afghan Restaurant  \
0                              Adelaide,King,Richmond           0.000000   
1                                         Berczy Park           0.000000   
2          Brockton,Exhibition Place,Parkdale Village           0.000000   
3   Business Reply Mail Processing Centre 969 Eastern           0.000000   
4   CN Tower,Bathurst Quay,Island airport,Harbourf...           0.000000   
5                          Cabbagetown,St. James Town           0.000000   
6                                  Central Bay Street           0.000000   
7             Chinatown,Grange Park,Kensington Market           0.000000   
8                                            Christie           0.000000   
9                                Church and Wellesley           0.011905   
10                      Commerce Court,Victoria Hotel           0.000000   
11                                         Davisville           0.000000   
12                                   Davisville North           0.000000   
13  Deer Park,Forest Hill SE,Rathnelly,South Hill,...           0.000000   
14            Design Exchange,Toronto Dominion Centre           0.000000   
15                        Dovercourt Village,Dufferin           0.000000   
16              First Canadian Place,Underground city           0.000000   
17                 Forest Hill North,Forest Hill West           0.000000   
18                      Harbord,University of Toronto           0.000000   
19                                       Harbourfront           0.000000   
20    Harbourfront East,Toronto Islands,Union Station           0.000000   
21                       High Park,The Junction South           0.000000   
22                                      Lawrence Park           0.000000   
23                            Little Portugal,Trinity           0.000000   
24                         Moore Park,Summerhill East           0.000000   
25                                 North Toronto West           0.000000   
26                              Parkdale,Roncesvalles           0.000000   
27                                           Rosedale           0.000000   
28                                           Roselawn           0.000000   
29                                  Runnymede,Swansea           0.000000   
30                            Ryerson,Garden District           0.000000   
31                                     St. James Town           0.000000   
32                    Stn A PO Boxes 25 The Esplanade           0.000000   
33                                    Studio District           0.000000   
34                  The Annex,North Midtown,Yorkville           0.000000   
35                                        The Beaches           0.000000   
36                      The Beaches West,India Bazaar           0.000000   
37                        The Danforth West,Riverdale           0.000000   

     Airport  Airport Food Court  Airport Lounge  Airport Service  \
0   0.000000            0.000000        0.000000         0.000000   
1   0.000000            0.000000        0.000000         0.000000   
2   0.000000            0.000000        0.000000         0.000000   
3   0.000000            0.000000        0.000000         0.000000   
4   0.071429            0.071429        0.142857         0.071429   
5   0.000000            0.000000        0.000000         0.000000   
6   0.000000            0.000000        0.000000         0.000000   
7   0.000000            0.000000        0.000000         0.000000   
8   0.000000            0.000000        0.000000         0.000000   
9   0.000000            0.000000        0.000000         0.000000   
10  0.000000            0.000000        0.000000         0.000000   
11  0.000000            0.000000        0.000000         0.000000   
12  0.000000            0.000000        0.000000         0.000000   
13  0.000000            0.000000        0.000000         0.000000   
14  0.000000            0.000000        

##### Let's print each neighborhood along with the top 5 most common venues

In [80]:
num_top_venues = 5

for hood in toronto_grouped['Neighbourhood']:
    print("----"+hood+"----")
    temp = toronto_grouped[toronto_grouped['Neighbourhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Adelaide,King,Richmond----
             venue  freq
0      Coffee Shop  0.07
1             Café  0.05
2       Steakhouse  0.04
3              Bar  0.04
4  Thai Restaurant  0.04


----Berczy Park----
                venue  freq
0         Coffee Shop  0.07
1              Bakery  0.05
2                Café  0.04
3        Cocktail Bar  0.04
4  Seafood Restaurant  0.04


----Brockton,Exhibition Place,Parkdale Village----
            venue  freq
0  Breakfast Spot  0.10
1            Café  0.10
2     Coffee Shop  0.10
3          Bakery  0.05
4   Grocery Store  0.05


----Business Reply Mail Processing Centre 969 Eastern----
                venue  freq
0  Light Rail Station  0.12
1         Yoga Studio  0.06
2                 Spa  0.06
3       Garden Center  0.06
4              Garden  0.06


----CN Tower,Bathurst Quay,Island airport,Harbourfront West,King and Spadina,Railway Lands,South Niagara----
              venue  freq
0    Airport Lounge  0.14
1  Airport Terminal  0.14
2          Bout

##### Let's put that into a *pandas* dataframe

In [81]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [82]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighbourhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighbourhood'] = toronto_grouped['Neighbourhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted

Neighbourhood 1st Most Common Venue  \
0                              Adelaide,King,Richmond           Coffee Shop   
1                                         Berczy Park           Coffee Shop   
2          Brockton,Exhibition Place,Parkdale Village           Coffee Shop   
3   Business Reply Mail Processing Centre 969 Eastern    Light Rail Station   
4   CN Tower,Bathurst Quay,Island airport,Harbourf...        Airport Lounge   
5                          Cabbagetown,St. James Town                  Park   
6                                  Central Bay Street           Coffee Shop   
7             Chinatown,Grange Park,Kensington Market                  Café   
8                                            Christie         Grocery Store   
9                                Church and Wellesley           Coffee Shop   
10                      Commerce Court,Victoria Hotel           Coffee Shop   
11                                         Davisville        Sandwich Place   
12                                   Davisville North          Dance Studio   
13  Deer Park,Forest Hill SE,Rathnelly,South Hill,...           Coffee Shop   
14            Design Exchange,Toronto Dominion Centre           Coffee Shop   
15                        Dovercourt Village,Dufferin              Pharmacy   
16              First Canadian Place,Underground city           Coffee Shop   
17                 Forest Hill North,Forest Hill West                  Park   
18                      Harbord,University of Toronto                  Café   
19                                       Harbourfront           Coffee Shop   
20    Harbourfront East,Toronto Islands,Union Station           Coffee Shop   
21                       High Park,The Junction South    Mexican Restaurant   
22                                      Lawrence Park                  Park   
23                            Little Portugal,Trinity                   Bar   
24                         Moore Park,Summerhill East            Playground   
25                                 North Toronto West        Clothing Store   
26                              Parkdale,Roncesvalles             Gift Shop   
27                                           Rosedale                  Park   
28                                           Roselawn                Garden   
29                                  Runnymede,Swansea           Coffee Shop   
30                            Ryerson,Garden District           Coffee Shop   
31                                     St. James Town                  Café   
32                    Stn A PO Boxes 25 The Esplanade           Coffee Shop   
33                                    Studio District                  Café   
34                  The Annex,North Midtown,Yorkville                  Café   
35                                        The Beaches     Health Food Store   
36                      The Beaches West,India Bazaar        Sandwich Place   
37                        The Danforth West,Riverdale      Greek Restaurant   

   2nd Most Common Venue 3rd Most Common Venue      4th Most Common Venue  \
0                   Café                   Bar            Thai Restaurant   
1                 Bakery          Cocktail Bar                   Beer Bar   
2                   Café        Breakfast Spot              Grocery Store   
3            Yoga Studio         Auto Workshop              Garden Center   
4       Airport Terminal                 Plane                        Bar   
5            Coffee Shop                Bakery                       Café   
6     Italian Restaurant          Burger Joint                       Café   
7                    Bar    Mexican Restaurant      Vietnamese Restaurant   
8                   Café                  Park          Convenience Store   
9    Japanese Restaurant      Sushi Restaurant                    Gay Bar   
10                  Café                 Hotel                 Restaurant   
11          Dessert Shop          

#### 3.6 Cluster Neighbourhoods

###### Run *k*-means to cluster the neighborhood into 5 clusters

In [83]:
# set number of clusters
kclusters = 5

toronto_grouped_clustering = toronto_grouped.drop('Neighbourhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10]

array([1, 1, 1, 1, 1, 1, 1, 1, 1, 1], dtype=int32)

In [84]:
toronto_merged = Toronto_data

# add clustering labels
toronto_merged['Cluster Labels'] = kmeans.labels_

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighbourhood'), on='Neighbourhood')

toronto_merged.head() # check the last columns!

PostalCode          Borough                  Neighbourhood   Latitude  \
0        M4E     East Toronto                    The Beaches  43.676357   
1        M4K     East Toronto    The Danforth West,Riverdale  43.679557   
2        M4L     East Toronto  The Beaches West,India Bazaar  43.668999   
3        M4M     East Toronto                Studio District  43.659526   
4        M4N  Central Toronto                  Lawrence Park  43.728020   

   Longitude  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
0 -79.293031               1     Health Food Store                 Trail   
1 -79.352188               1      Greek Restaurant           Coffee Shop   
2 -79.315572               1        Sandwich Place            Board Shop   
3 -79.340923               1                  Café           Coffee Shop   
4 -79.388790               1                  Park           Swim School   

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
0                   Pub          Dessert Shop    Falafel Restaurant   
1    Italian Restaurant        Ice Cream Shop            Restaurant   
2     Food & Drink Shop    Italian Restaurant      Sushi Restaurant   
3    Italian Restaurant   American Restaurant                Bakery   
4              Bus Line           Yoga Studio                 Diner   

  6th Most Common Venue   7th Most Common Venue    8th Most Common Venue  \
0           Event Space    Ethiopian Restaurant        Electronics Store   
1             Bookstore  Furniture / Home Store        Indian Restaurant   
2     Fish & Chips Shop                     Pub           Ice Cream Shop   
3        Clothing Store                   Diner  Comfort Food Restaurant   
4    Falafel Restaurant             Event Space     Ethiopian Restaurant   

         9th Most Common Venue       10th Most Common Venue  
0  Eastern European Restaurant          Dumpling Restaurant  
1                    Juice Bar                 Liquor Store  
2                Movie Theater         Fast Food Restaurant  
3                  Cheese Shop    Middle Eastern Restaurant  
4            Electronics Store  Eastern European Restaurant

#### 3.7 Final resulting clusters visualization - Toronto Neighborhoods

In [85]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i+x+(i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged['Latitude'], toronto_merged['Longitude'], toronto_merged['Neighbourhood'], toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

#### 3.8 Examine Clusters

##### Cluster 1:

In [86]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 0, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
24  Central Toronto               0                  Café   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
24        Sandwich Place           Coffee Shop              Pharmacy   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
24             BBQ Joint        Cosmetics Shop           Pizza Place   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
24                   Pub         Metro Station           Burger Joint

##### Cluster 2:

In [87]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 1, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
0       East Toronto               1     Health Food Store   
1       East Toronto               1      Greek Restaurant   
2       East Toronto               1        Sandwich Place   
3       East Toronto               1                  Café   
4    Central Toronto               1                  Park   
5    Central Toronto               1          Dance Studio   
6    Central Toronto               1        Clothing Store   
7    Central Toronto               1        Sandwich Place   
8    Central Toronto               1            Playground   
9    Central Toronto               1           Coffee Shop   
10  Downtown Toronto               1                  Park   
11  Downtown Toronto               1                  Park   
12  Downtown Toronto               1           Coffee Shop   
13  Downtown Toronto               1           Coffee Shop   
14  Downtown Toronto               1           Coffee Shop   
15  Downtown Toronto               1                  Café   
16  Downtown Toronto               1           Coffee Shop   
18  Downtown Toronto               1           Coffee Shop   
19  Downtown Toronto               1           Coffee Shop   
20  Downtown Toronto               1           Coffee Shop   
21  Downtown Toronto               1           Coffee Shop   
23   Central Toronto               1                  Park   
25  Downtown Toronto               1                  Café   
26  Downtown Toronto               1                  Café   
29  Downtown Toronto               1           Coffee Shop   
30  Downtown Toronto               1         Grocery Store   
31      West Toronto               1              Pharmacy   
32      West Toronto               1                   Bar   
33      West Toronto               1           Coffee Shop   
34      West Toronto               1    Mexican Restaurant   
35      West Toronto               1             Gift Shop   
36      West Toronto               1           Coffee Shop   
37      East Toronto               1    Light Rail Station   

   2nd Most Common Venue 3rd Most Common Venue      4th Most Common Venue  \
0                  Trail                   Pub               Dessert Shop   
1            Coffee Shop    Italian Restaurant             Ice Cream Shop   
2             Board Shop     Food & Drink Shop         Italian Restaurant   
3            Coffee Shop    Italian Restaurant        American Restaurant   
4            Swim School              Bus Line                Yoga Studio   
5                    Gym        Breakfast Spot                       Park   
6    Sporting Goods Shop           Coffee Shop                Yoga Studio   
7           Dessert Shop           Pizza Place           Sushi Restaurant   
8            Summer Camp            Restaurant                Yoga Studio   
9                    Pub           Pizza Place                 Sports Bar   
10            Playground                 Trail                Yoga Studio   
11           Coffee Shop                Bakery                       Café   
12   Japanese Restaurant      Sushi Restaurant                    Gay Bar   
13                   Pub                  Park                     Bakery   
14        Clothing Store                  Café  Middle Eastern Restaurant   
15            Restaurant           Coffee Shop         Italian Restaurant   
16                Bakery          Cocktail Bar                   Beer Bar   
18                  Café                   Bar            Thai Restaurant   
19                 Hotel              Aquarium                       Café   
20                  Café                 Hotel                        Bar   
21                  Café                 Hotel                 Restaurant   
23                 Trail         Jewelry Store           Sushi Restaurant   
25             Bookstore        Sandwich Place                 Restaurant   
26                   Bar    Mexican Restaurant      Viet

##### Cluster 3:

In [88]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 2, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
28  Downtown Toronto               2           Coffee Shop   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
28                  Café            Restaurant                Bakery   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
28    Italian Restaurant    Seafood Restaurant              Beer Bar   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
28   Japanese Restaurant                 Hotel            Cheese Shop

##### Cluster 4:

In [89]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 3, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
17  Downtown Toronto               3           Coffee Shop   
27  Downtown Toronto               3        Airport Lounge   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
17    Italian Restaurant          Burger Joint                  Café   
27      Airport Terminal                 Plane                   Bar   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
17        Ice Cream Shop        Sandwich Place  Gym / Fitness Center   
27           Coffee Shop      Sculpture Garden              Boutique   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
17       Bubble Tea Shop           Salad Place                    Spa  
27         Boat or Ferry       Harbor / Marina        Airport Service

##### Cluster 5:

In [90]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 4, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
22  Central Toronto               4                Garden   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
22           Yoga Studio                 Diner        Farmers Market   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
22    Falafel Restaurant           Event Space  Ethiopian Restaurant   

   8th Most Common Venue        9th Most Common Venue 10th Most Common Venue  
22     Electronics Store  Eastern European Restaurant    Dumpling Restaurant